src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -57,19 +111,106 @@ internal ConfigurationClientManager(IList<ConfigurationClientWrapper> clients)
             _clients = clients;
         }
 
-        public IEnumerable<ConfigurationClient> GetAvailableClients(DateTimeOffset time)
+        public async IAsyncEnumerable<ConfigurationClient> GetAvailableClients([EnumeratorCancellation] CancellationToken cancellationToken)
         {
-            return _clients.Where(client => client.BackoffEndTime <= time).Select(c => c.Client).ToList();
-        }
+            DateTimeOffset now = DateTimeOffset.UtcNow;
+            DateTimeOffset minimalRefreshTime = _lastFallbackClientRefreshAttempt + MinimalClientRefreshInterval;
+            Task task = null;
+
+            // Principle of refreshing fallback clients:
+            // 1.Perform initial refresh attempt to query available clients on app start-up.
+            // 2.Refreshing when cached fallback clients have expired, FallbackClientRefreshExpireInterval is due.
+            // 3.At least wait MinimalClientRefreshInterval between two attempt to ensure not perform refreshing too often.
+
+            if (_replicaDiscoveryEnabled &&
+                now >= minimalRefreshTime && 
+                (_dynamicClients == null ||
+                now >= _lastFallbackClientRefresh + FallbackClientRefreshExpireInterval))
+            {
+                _lastFallbackClientRefreshAttempt = now;
+
+                var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);

```

Comment:
Is it necessary to tie it to the caller cancellation? Caller cancellation could be canceled quite fast and never give a chance for this background operation to complete. Seems like it may deserve it's own decoupled cancellation.